src/cli/cli_uart.cpp

```diff
@@ -237,25 +237,52 @@ otError Uart::ProcessCommand(void)
 int Uart::Output(const char *aBuf, uint16_t aBufLength)
 {
     OT_CLI_UART_OUTPUT_LOCK();
-    uint16_t remaining = kTxBufferSize - mTxLength;
-    uint16_t tail;
+    uint16_t sent = 0;
 
-    if (aBufLength > remaining)
+    while (aBufLength > 0)
     {
-        aBufLength = remaining;
-    }
+        uint16_t remaining = kTxBufferSize - mTxLength;
+        uint16_t tail;
+        uint16_t sendLength = aBufLength;
 
-    for (int i = 0; i < aBufLength; i++)
-    {
-        tail            = (mTxHead + mTxLength) % kTxBufferSize;
-        mTxBuffer[tail] = *aBuf++;
-        mTxLength++;
+        if (sendLength > remaining)
+        {
+            sendLength = remaining;
+        }
+
+        for (int i = 0; i < sendLength; i++)
+        {
+            tail            = (mTxHead + mTxLength) % kTxBufferSize;
+            mTxBuffer[tail] = *aBuf++;
+            aBufLength--;
+            mTxLength++;
+        }
+
+        Send();
+
+        sent += sendLength;
+
+        if (aBufLength > 0)
+        {
+            // More to send, so flush what's waiting now
+            otError err = otPlatUartFlush();
+
+            if (err == OT_ERROR_NONE)
+            {
+                // Flush successful, reset the pointers
+                SendDoneTask();
```

Comment:
There might be a reentrant problem of `Send()`.

`SendDoneTask()` may call `Send()` if previous requested data is not fully delivered to platform layer, because requested data might be delivered to platform layer twice for the ring buffer design.  As a result, part of data will be lost.

I think we should check if `mTxLength` is `0` after this call, and if so, call `otPlatUartFlush()` again.

Reply:
Not sure how it'd be re-entrant, the flush operation is not meant to call `otPlatUartSendDone` itself, as this is implied by the flush operation (it's either meant to wait until it is done then return no error, or return an error having failed to flush).

`otPlatUartSendDone` should only be called if and only if, after `Output` has finished, there is left-over data to be processed by the driver at its leisure, and the signal indicates *that* data has been sent.

We could split `SendDoneTask()` up I guess, make a `MarkPendingSent()` function (or any other name) which would be:
```
void Uart::MarkPendingSent(void)
{
    mTxHead = (mTxHead + mSendLength) % kTxBufferSize;
    mTxLength -= mSendLength;
    mSendLength = 0;
}
```

We would call *that* instead in `Output`, and `SendDoneTask` would be simplified to:

```
void Uart::SendDoneTask(void)
{
    MarkPendingSent();
    Send();
}
```

The C++ compiler should optimise this to remove the function call overhead.

Reply:
Let's say before line `Send():261`, `mTxHead > 0`, `mTxLength == sizeof(mTxBuffer)`, after `Send()`, `mSendLength == kTxBufferSize - mTxHead`. 

After `otPlatUartFlush():268`, `kTxBufferSize - mTxHead` bytes are sent, but `mTxHead` bytes are left.

After `SendDoneTask():273`, `mSendLength == 0`, but `mTxLength == mTxHead`, and the left bytes are delivered to platform, but not actually sent.

Then more bytes are filled and it goes to `Send():261` again, however, this time, since `mSendLength` is not `0`, it directly returns and only `mTxHead` bytes get flushed by `otPlatUartFlush():268`. Those newly appended byte are not sent.

Reply:
Okay… so I did a bit of pen-and-paper analysis this afternoon, it's very possible I missed something, but after analysing what I understand the code to be doing, I still don't see the data loss issue.

I will apologise right up front about my phone having the world's worst auto-focus algorithm (maybe ZTE should patent it… the "auto-defocus" algorithm).

![send-analysis-01](https://user-images.githubusercontent.com/1710526/61612437-ee587f00-aca1-11e9-846d-4d3143b19a28.png)
![send-analysis-02](https://user-images.githubusercontent.com/1710526/61612438-eef11580-aca1-11e9-94a2-3328db918c66.png)
![send-analysis-03](https://user-images.githubusercontent.com/1710526/61612439-eef11580-aca1-11e9-89d3-2a0c8ec76ca0.png)

On paper, it looks to be doing exactly what it says on the tin.

Reply:
I think you are right! Thank you!